diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc
index cb7231c20..7d658b443 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc
@@ -1531,6 +1531,9 @@ bool PrintC::pushPtrCharConstant(uintb val,const TypePointer *ct,const Varnode *
 {
   if (val==0) return false;
   AddrSpace *spc = glb->getDefaultDataSpace();
+  if (!vn->hasNoDescend()) {
+      spc = (*vn->beginDescend())[0].getParent()->getFuncdata()->getAddress().getSpace();
+  }
   uintb fullEncoding;
   Address stringaddr = glb->resolveConstant(spc,val,ct->getSize(),op->getAddr(),fullEncoding);
   if (stringaddr.isInvalid()) return false;
@@ -1560,6 +1563,9 @@ bool PrintC::pushPtrCodeConstant(uintb val,const TypePointer *ct,
 				    const PcodeOp *op)
 {
   AddrSpace *spc = glb->getDefaultCodeSpace();
+  if (!vn->hasNoDescend()) {
+      spc = (*vn->beginDescend())[0].getParent()->getFuncdata()->getAddress().getSpace();
+  }
   Funcdata *fd = (Funcdata *)0;
   val = AddrSpace::addressToByte(val,spc->getWordSize());
   fd = glb->symboltab->getGlobalScope()->queryFunction( Address(spc,val));
